Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/GraphicsManagement.h>
+#include <Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
+namespace Kernel::Graphics {
+
+unsigned VirtIOFrameBufferDevice::next_device_id = 0;
+
+VirtIOFrameBufferDevice::VirtIOFrameBufferDevice(RefPtr<VirtIOGPU> virtio_gpu)
+    : BlockDevice(29, GraphicsManagement::the().allocate_minor_device_number())
+    , m_gpu(virtio_gpu)
+{
+    auto write_sink_page = MM.allocate_user_physical_page(MemoryManager::ShouldZeroFill::No).release_nonnull();
+    auto num_needed_pages = m_gpu->framebuffer_vm_object().page_count();
+    NonnullRefPtrVector<PhysicalPage> pages;
+    for (auto i = 0u; i < num_needed_pages; ++i) {
+        pages.append(write_sink_page);
+    }
+    m_framebuffer_sink_vmobject = adopt_ref(*new AnonymousVMObject(move(pages)));
+}
+
+VirtIOFrameBufferDevice::~VirtIOFrameBufferDevice()
+{
+}
+
+int VirtIOFrameBufferDevice::ioctl(FileDescription&, unsigned request, FlatPtr arg)
+{
+    REQUIRE_PROMISE(video);
+    switch (request) {
+    case FB_IOCTL_GET_SIZE_IN_BYTES: {
+        auto* out = (size_t*)arg;
+        size_t value = m_gpu->framebuffer_size_in_bytes();
+        if (!copy_to_user(out, &value))
+            return -EFAULT;
+        return 0;
+    }
+    case FB_IOCTL_GET_RESOLUTION:
+    case FB_IOCTL_SET_RESOLUTION: {
```

Comment:
is GET and SET resolution supposed to be the same thing?

Reply:
Oh, so virtiogpu currently don't support setting the resolution. I guess I'm going to have to implement this before merging, it would be bad to regress to not being able to set it.

Reply:
I've now added resolution setting.